### ISSUE TITLE
v2ray plugin is failing due to upstream server issue

### ISSUE THREAD

**Issue by nickgnazzo** [StreisandEffect/streisand#1642]
Just posting this here for visibility in case anyone trying to use the v2ray plugin is encountering this issue. If you enable the v2ray plugin for Shadowsocks for your Streisand server, the install will likely fail with an "insecure redirect" error. This is because of what appears to be a misconfiguration on the server or service hosting v2ray.com.

Edit: I believe this is also what's causing all the current CI jobs to fail. The Travis CI job logs show the same error reported below.

I posted an issue on the v2ray github (which I hope was the right place, supposedly it's the source code for v2ray.com) reporting the details. Not sure if the owner will see and be able to fix but I will also try sending an email to the v2ray server admin if this gets ignored:

https://github.com/v2ray/manual/issues/278

### Expected behavior:

v2ray plugin will get installed and I can revel in the glory of ansible provisioning.

### Actual Behavior:

v2ray plugin does not get installed and I cry a little.

Error below:

>TASK [shadowsocks : Get V2Ray-plugin] > ********************************************************************************************************
> fatal: [x.x.x.x]: FAILED! => {"changed": true, "cmd": "go get github.com/shadowsocks/v2ray-plugin", "delta": "0:00:02.478312", "end": "2019-09-07 19:38:15.536559", "msg": "non-zero return code", "rc": 2, "start": "2019-09-07 19:38:13.058247", "stderr": "package v2ray.com/core: unrecognized import path \"v2ray.com/core\" (https fetch: Get http://www.v2ray.com/core/?go-get=1: redirected from secure URL https://v2ray.com/core?go-get=1 to insecure URL http://www.v2ray.com/core/?go-get=1)\npanic: runtime error: invalid memory address or nil pointer dereference\n[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x889fad]\n\ngoroutine 1 [running]:\ncmd/go/internal/get.metaImportsForPrefix.func2(0x0, 0x0, 0xb27d60, 0xc00098ce70)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:907 +0x29d\ninternal/singleflight.(*Group).doCall(0xeabf80, 0xc0002c4140, 0xc000205e40, 0xe, 0xc0001892f8)\n\t/usr/lib/go-1.13/src/internal/singleflight/singleflight.go:95 +0x2e\ninternal/singleflight.(*Group).Do(0xeabf80, 0xc000205e40, 0xe, 0xc0001892f8, 0x0, 0xc00098cbd0, 0x0, 0x0, 0xc000025201)\n\t/usr/lib/go-1.13/src/internal/singleflight/singleflight.go:63 +0x1ba\ncmd/go/internal/get.metaImportsForPrefix(0xc000205e40, 0xe, 0x0, 0x0, 0x1d, 0xc000205e40, 0xe, 0xc000205e4f, 0x3, 0xc000205e53)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:893 +0xf1\ncmd/go/internal/get.repoRootForImportDynamic(0xc000025201, 0x1d, 0x0, 0x0, 0x0, 0x0, 0x0)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:827 +0x665\ncmd/go/internal/get.RepoRootForImportPath(0xc000025201, 0x1d, 0x0, 0x0, 0xffffffffffffffff, 0xc0001896c0, 0xc000189838)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:662 +0x345\ncmd/go/internal/get.downloadPackage(0xc0001f8480, 0xc0000e3200, 0xc000025201)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:448 +0x1568\ncmd/go/internal/get.download(0xc000025201, 0x1d, 0xc000098d80, 0xc000189d78, 0x0)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:275 +0xd26\ncmd/go/internal/get.download(0x7ffc1de4be73, 0x23, 0x0, 0xc000189d78, 0x0)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:371 +0x718\ncmd/go/internal/get.runGet(0xea3a60, 0xc0000201a0, 0x1, 0x1)\n\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:162 +0x170\nmain.main()\n\t/usr/lib/go-1.13/src/cmd/go/main.go:189 +0x57f", "stderr_lines": ["package v2ray.com/core: unrecognized import path \"v2ray.com/core\" (https fetch: Get http://www.v2ray.com/core/?go-get=1: redirected from secure URL https://v2ray.com/core?go-get=1 to insecure URL http://www.v2ray.com/core/?go-get=1)", "panic: runtime error: invalid memory address or nil pointer dereference", "[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x889fad]", "", "goroutine 1 [running]:", "cmd/go/internal/get.metaImportsForPrefix.func2(0x0, 0x0, 0xb27d60, 0xc00098ce70)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:907 +0x29d", "internal/singleflight.(*Group).doCall(0xeabf80, 0xc0002c4140, 0xc000205e40, 0xe, 0xc0001892f8)", "\t/usr/lib/go-1.13/src/internal/singleflight/singleflight.go:95 +0x2e", "internal/singleflight.(*Group).Do(0xeabf80, 0xc000205e40, 0xe, 0xc0001892f8, 0x0, 0xc00098cbd0, 0x0, 0x0, 0xc000025201)", "\t/usr/lib/go-1.13/src/internal/singleflight/singleflight.go:63 +0x1ba", "cmd/go/internal/get.metaImportsForPrefix(0xc000205e40, 0xe, 0x0, 0x0, 0x1d, 0xc000205e40, 0xe, 0xc000205e4f, 0x3, 0xc000205e53)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:893 +0xf1", "cmd/go/internal/get.repoRootForImportDynamic(0xc000025201, 0x1d, 0x0, 0x0, 0x0, 0x0, 0x0)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:827 +0x665", "cmd/go/internal/get.RepoRootForImportPath(0xc000025201, 0x1d, 0x0, 0x0, 0xffffffffffffffff, 0xc0001896c0, 0xc000189838)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/vcs.go:662 +0x345", "cmd/go/internal/get.downloadPackage(0xc0001f8480, 0xc0000e3200, 0xc000025201)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:448 +0x1568", "cmd/go/internal/get.download(0xc000025201, 0x1d, 0xc000098d80, 0xc000189d78, 0x0)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:275 +0xd26", "cmd/go/internal/get.download(0x7ffc1de4be73, 0x23, 0x0, 0xc000189d78, 0x0)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:371 +0x718", "cmd/go/internal/get.runGet(0xea3a60, 0xc0000201a0, 0x1, 0x1)", "\t/usr/lib/go-1.13/src/cmd/go/internal/get/get.go:162 +0x170", "main.main()", "\t/usr/lib/go-1.13/src/cmd/go/main.go:189 +0x57f"], "stdout": "", "stdout_lines": []}

### Steps to Reproduce:
1. Run Streisand with Shadowsocks and v2ray plugin enabled

[ contents of `streisand-diagnostics.md` here ]

**Comment by nickgnazzo** [comment]
So it looks like this is the same behavior encountered by some other go projects hosted on GitHub Pages. There's a bug causing GitHub Pages to drop the https protocol when sending the HTTP 301 response (the `location` header strips away https and sends back http):

https://github.com/golang/go/issues/32071

The GitHub team seem to have deployed a fix that worked for googlemaps.github.io, but the issue is still present for v2ray.github.io and v2ray.com:

```
[user@personal ~]$ curl -si https://v2ray.com/core?go-get=1 | grep -i location
location: http://www.v2ray.com/core/?go-get=1
[user@personal ~]$ curl -si https://v2ray.github.io/core?go-get=1 | grep -i location
location: http://www.v2ray.com/core?go-get=1
```

Appending a trailing slash, at least for the custom domain (`v2ray.com`) "fixes" the issue. The location header will have the correct protocol:

```
[user@personal ~]$ curl -si https://v2ray.com/core/?go-get=1 | grep -i location
location: https://www.v2ray.com/core/index.html?go-get=1
```

However, that doesn't work if you try the same thing with the github.io domain:

```
[user@personal ~]$ curl -si https://v2ray.github.io/core/?go-get=1 | grep -i location
location: http://www.v2ray.com/core/?go-get=1

```

Anyway, for whatever reason, the fix deployed by GitHub doesn't seem to have fixed the v2ray GitHub Pages site. That being said, the golang/go team appears to have created a workaround for this issue where they ignore the incorrect location header in 301 responses (unless you use `-insecure`), and keep the protocol as https when doing `go get` for URLs that have this issue, as documented here:

https://github.com/golang/go/issues/34049

(You can also see them get the same exact issue when trying `go get` for various v2ray.com packages).

According to that same thread (comment [here](https://github.com/golang/go/issues/34049#issuecomment-527964010)) they backported the workaround to 1.12.9, and 1.13--which seems like the default version when Streisand runs the golang install via apt.

It looks like the golang package installed by Streisand is coming from an Ubuntu PPA [over here](https://launchpad.net/~longsleep/+archive/ubuntu/golang-backports/+packages). I believe though that the last successful build of the golang packages in that PPA might have come _before_ the workaround patch for this https thing. If you look at the last successful build for golang-1.13 amd64 within longsleep's PPA:

https://launchpad.net/~longsleep/+archive/ubuntu/golang-backports/+build/17517913

It seems like it was done around the same time the backport was being committed, so my guess is the build missed the window there and just needs to be rebuilt in order to get that workaround. Not knowing much about how golang does its releases or publishes them to distros like Ubuntu/Debian/etc, this is just a wild guess on my part. I'll follow-up with the maintainer of the PPA and see what he says.

In any event, you can still confirm the fix works in a newer distro like fedora 30, even with golang 1.12.9:

```
[user@disp2972 ~]$ cat /etc/fedora-release 
Fedora release 30 (Thirty)
[user@disp2972 ~]$ go get v2ray.com/core
[user@disp2972 ~]$ ls go/src/v2ray.com/core/
LICENSE                       context.go           main
README.md                     context_test.go      mocks.go
WORKSPACE                     core.go              proto.go
annotations.go                errors.generated.go  proxy
app                           external             release
azure-pipelines.template.yml  features             testing
azure-pipelines.yml           functions.go         transport
common                        functions_test.go    v2ray.go
config.go                     go.mod               v2ray_test.go
config.pb.go                  go.sum
config.proto                  infra
[user@disp2972 ~]$ go version
go version go1.12.9 linux/amd64
```

**Comment by sysk** [comment]
Is there a workaround for this? I need to install streisand+ v2ray asap :(

**Comment by nickgnazzo** [comment]
Yeah, if you're able to start a new Streisand server from scratch, you can use this small patch I wrote to your local copy of the Streisand repo, which makes Streisand manually clone the v2ray-core repo to avoid this error:

1. `cd /path/to/Streisand/repo`
2. `curl -O https://gist.githubusercontent.com/nickgnazzo/cb424f3b94568431ccd07409eeeb8f20/raw/0ddb8bfe665e06a22c8c149fe17b9e393e8f8880/0001-Temporary-workaround-for-v2ray-core-insecure-redirec.patch`
3. `git am < 0001-Temporary-workaround-for-v2ray-core-insecure-redirec.patch`
4. Re-run Streisand script with shadowsocks and v2ray plugin enabled.


I tested this with shadowsocks and the v2ray plugin enabled and it seemed to work. Let me know how it goes for you. I might open a pull request to merge this into master if it seems like a reasonable temporary workaround to the maintainers?

However, if you already have a Streisand server and can't create a new one, and you enabled the shadowsocks role (but left v2ray disabled), you can install the v2ray plugin by manually cloning the v2ray/core dependency to your `GOPATH`:

First setup your GOPATH if it isn't already, then clone the repo. Note the destination path at the end of the `git clone` command:

```
root@streisand:~# GOPATH="${GOPATH:=$HOME/go}"
root@streisand:~# mkdir -p $GOPATH/src/v2ray.com/
root@streisand:~# git clone https://github.com/v2ray/v2ray-core $GOPATH/src/v2ray.com/core
Cloning into '/root/go/src/v2ray.com/core'...
remote: Enumerating objects: 40626, done.
remote: Total 40626 (delta 0), reused 0 (delta 0), pack-reused 40626
Receiving objects: 100% (40626/40626), 28.48 MiB | 16.70 MiB/s, done.
Resolving deltas: 100% (28674/28674), done.
Checking connectivity... done.
```

Now you should be able to run `go get` without the redirect error:

```
root@streisand:~# go get github.com/shadowsocks/v2ray-plugin
```

Copy the v2ray-plugin binary to shadowsocks-libev location, then modify your shadowsocks config to use the plugin and cover domain (default for Streisand is `github.com`): 

```
root@streisand:~# cp -rf /root/go/bin/v2ray-plugin /etc/shadowsocks-libev
root@streisand:~# vim /etc/shadowsocks-libev/config.json
```

Add these two lines at the end of your config.json, before the ending curly brace. Also make sure you add a trailing comma to the option just above your new ones, or the JSON will fail to validate:

```
"plugin":"/etc/shadowsocks-libev/v2ray-plugin",
"plugin_opts":"server;host=github.com"
```
In other words, make sure `/etc/shadowsocks-libev/config.json` looks something like this (notice the trailing comma at the end of the "fast_open" line):

```
{
    "server":"x.x.x.x",
    "server_port":xxxx,
    "local_port":xxxx,
    "password":"xxxxxxxxx",
    "timeout":600,
    "method":"chacha20-ietf-poly1305",
    "fast_open":true,
    "plugin":"/etc/shadowsocks-libev/v2ray-plugin",
    "plugin_opts":"server;host=github.com"
}
```

Now, restart your shadowsocks service

```
root@streisand:~# systemctl restart shadowsocks-libev.service
```

Only bad thing about this option is that your Streisand gateway page won't have up-to-date instructions for how to connect with v2ray.
